utils/log_parser.py

```diff
@@ -0,0 +1,303 @@
+
+    def get_color(self, node):
+        idx = hash(node.get_kind_name()) % len(self.colors)
+        return self.colors[idx]
+
+    def dump_node(self, node):
+        if not node:
+            return
+        nodeStr = self.get_unique_vertex_name(node) + "[\n"
+        nodeStr += "\tlabel = \""
+        nodeStr += self.dump_label(node)
+        nodeStr += "\"\n"
+        nodeStr += "\tshape = \"record\"\n"
+        nodeStr += "\tstyle=\"filled,rounded\"\n"
+        nodeStr += "\tfillcolor="
+        nodeStr += self.get_color(node)
+        nodeStr += "\npenwidth = 2];\n"
+        self.vertices_.append(nodeStr)
+
+    def visitNodes(self):
+        for k, n in NODES_MAP:
+            self.dump_node(NODES_MAP[(k, n)])
+
+    def visitEdges(self):
+        for k, n in NODES_MAP:
+            node = NODES_MAP[(k, n)]
+            for nodeInput in node.get_inputs():
+                i = nodeInput[0]
+                if (i.get_kind_name(), i.get_name()) not in NODES_MAP:
+                    print(i.get_kind_name(), i.get_name())
+                edgeStr = self.get_unique_vertex_name(i)+":Outputs -> "
+                edgeStr += self.get_unique_vertex_name(node)+":Inputs"
+                self.edges_.append(edgeStr)
+
+    def dump_graph(self, dagName):
+        self.visitNodes()
+        self.visitEdges()
+        f = open(dagName + "_dotty.dot", "w")
+        f.write("digraph DAG {\n\trankdir=TB;\n")
+        for v in self.vertices_:
+            f.write(v + "\n")
+        for e in self.edges_:
+            f.write(e + ";\n")
+        f.write("}")
+        f.close()
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Parse compilation log")
+    parser.add_argument("-f", "--log-file")
+    parser.add_argument("--dump-phases", nargs="+")
+    options = parser.parse_args()
+
+    if options.dump_phases:
+        dumpPhases = options.dump_phases
+    else:
+        dumpPhases = []
+
+    return options.log_file, dumpPhases
+
+
+def dump_dag(dagName):
+    dotty = DottyPrinter()
+    dotty.dump_graph(dagName)
+
+
+def process(logLines, dumpPhases):
+    phaseNo = 0
+    pCreate = re.compile(
+        '^.*\[FULL.*SCOPE:(.*)\].*CREATE.*\{.*\(Kind:(.*),.*Name:(.*)\).*<==(.*)\}$')
+    pInputs = re.compile('\([^\(\)]*\)')
+    pOneInput = re.compile('^\(Kind:(.*),.*Name:(.*),.*ResNo:(.*)\)$')
+    pChange = re.compile(
+        '^.*\[FULL.*SCOPE:(.*)\].*NODE_INPUT_CHANGE.*\{.*User\(Kind:(.*),.*Name:(.*)\).*::.*PrevOprValue(.*)->.*NewOprValue(.*)\}$')
+    pOpr = re.compile('^\(.*Kind:(.*),.*Name:(.*),.*ResNo:(.*)\)$')
+    pDelete = re.compile(
+        '^.*\[FULL.*SCOPE:(.*)\].*DELETE.*\(Kind:(.*),.*Name:(.*)\).*}$')
+    pEnter = re.compile('^ENTERSCOPE:(.*)$')
+    pExit = re.compile('^EXITSCOPE:(.*)$')
+
+    for ln in logLines:
+        # Process CREATE statement
+        if "CREATE" in ln:
+            m = re.match(pCreate, ln)
+            if m:
+                g = m.groups()
+                assert len(g) == 4
```

Comment:
Consider adding a helpful message here in case this `assert` fails (and for all others in the file).